### PR TITLE
Move detach to variable

### PR DESCRIPTION
**Pull request by ebetica**
I'm using the C++ Autograd API, and need a detach in C++. Tests seem to pass.

### CODE REVIEW

File: torch/csrc/autograd/variable.h [hunk 1]
```diff
@@ -275,6 +278,21 @@ inline bool& Variable::is_volatile() {
   return get()->is_volatile;
 }
 
+inline Variable Variable::detach() const {
```
Comment (colesbury) [inline torch/csrc/autograd/variable.h]: Can you move these definitions to `variable.cpp`? (And remove the inline flag). I'm worried that as variable.h gets bigger compilation time will slow down since it's used everywhere.

File: tools/autograd/templates/python_variable_methods.cpp [hunk 2]
```diff
@@ -20,23 +20,14 @@ static PyObject * THPVariable_detach(PyObject* self, PyObject* args)
 {
   HANDLE_TH_ERRORS
   auto& self_ = reinterpret_cast<THPVariable*>(self)->cdata;
-  Variable detached = make_variable(self_.data());
-  detached.is_volatile() = self_.is_volatile();
-  detached.version_counter() = self_.version_counter();
-  return THPVariable_Wrap(std::move(detached));
+  return THPVariable_Wrap(std::move(self_.detach()));
```
Comment (colesbury) [inline tools/autograd/templates/python_variable_methods.cpp]: No need for a std::move here.

**Review by colesbury — COMMENTED** [review]
two small comments. lgtm!

**Review by colesbury — APPROVED** [review]
lgtm

**Comment by soumith** [comment]
thanks @ebetica !
